File: language/runtime/apple/core/Sources/RuntimeHostAppleCore/Module/Location/LocationRequests.swift
import Foundation

/// The location request surface attached to one Apple runtime host.
@MainActor
public protocol LocationRequests {
    /// Read whether Apple location services are enabled.
    func locationServicesEnabled() -> RuntimeHostLocationServicesResponse

    /// Read one last-known Apple location sample.
    func locationLastKnown() -> RuntimeHostLocationLastKnownResponse

    /// Open one Apple location watch.
    func locationWatchOpen(
        watchID: String,
        options: RuntimeHostLocationWatchOptions
    ) -> UInt32

    /// Close one Apple location watch.
    func locationWatchClose(
        watchID: String
    ) -> UInt32
}

/// The explicit unsupported location request surface for one Apple runtime host.
@MainActor
public final class UnsupportedLocationRequests: LocationRequests {
    /// Create one unsupported location request surface.
    public init() {}

    public func locationServicesEnabled() -> RuntimeHostLocationServicesResponse {
        RuntimeHostLocationServicesResponse(status: 1)
    }

    public func locationLastKnown() -> RuntimeHostLocationLastKnownResponse {
        RuntimeHostLocationLastKnownResponse(status: 1)
    }

    public func locationWatchOpen(
        watchID: String,
        options: RuntimeHostLocationWatchOptions
    ) -> UInt32 {
        let _ = watchID
        let _ = options

        return 1
    }

    public func locationWatchClose(
        watchID: String
    ) -> UInt32 {
        let _ = watchID

        return 1
    }
}

File: language/runtime/android/kotlin/src/main/kotlin/dev/destack/runtime/android/module/location/LocationRequests.kt
package dev.destack.runtime.android.module.location

import dev.destack.runtime.android.core.hostStatusNotSupported

/**
 * The location request surface attached to one Android runtime host.
 */
public interface LocationRequests {
    /**
     * Read whether Android location services are enabled.
     */
    public fun locationServicesEnabled(): RuntimeHostLocationServicesResponse

    /**
     * Read one last-known Android location sample.
     */
    public fun locationLastKnown(): RuntimeHostLocationLastKnownResponse

    /**
     * Open one Android location watch.
     */
    public fun locationWatchOpen(
        watchId: String,
        options: RuntimeHostLocationWatchOptions,
    ): Int

    /**
     * Close one Android location watch.
     */
    public fun locationWatchClose(
        watchId: String,
    ): Int
}

/**
 * The explicit unsupported location request surface for one Android runtime host.
 */
public object UnsupportedLocationRequests : LocationRequests {
    override fun locationServicesEnabled(): RuntimeHostLocationServicesResponse {
        return RuntimeHostLocationServicesResponse(status = hostStatusNotSupported)
    }

    override fun locationLastKnown(): RuntimeHostLocationLastKnownResponse {
        return RuntimeHostLocationLastKnownResponse(status = hostStatusNotSupported)
    }

    override fun locationWatchOpen(
        watchId: String,
        options: RuntimeHostLocationWatchOptions,
    ): Int {
        return hostStatusNotSupported
    }

    override fun locationWatchClose(
        watchId: String,
    ): Int {
        return hostStatusNotSupported
    }
}
